tripal/tripal
# Improve maintainability chado upgrader complexity

# Tripal 4 Core Dev Task

### Issue #1404 


### Tripal Version: 4

## Description


This PR cleans-up the upgrader task to better meet CodeClimate expectations. I'm working to get PR's up for any work remaining in old branches where applicable. This was done when I was trying to get us to better maintainability once codeclimate checking was added.

## Testing?



### Automated testing

This task does have an automated test that performs the upgrade of a test schema to the Chado 1.3 specification. It tests by 
1. Creating the a base test chado
2. Removing the project.description column, adding an extra column and dropping a table.
3. Running the upgrader task and outputting the commands to a file.
4. Checking that the task completed successfully.

NOTE: This test is not very good since the upgrader task appears to be broken in 4.x and in this PR when applied directly 😢 

### Manual Testing

We are not testing that the upgrader works since it doesn't on either 4.x or this branch 😢 

Rather we are testing that this PR did not make anything worse 🙈 

1. Setup a fresh docker container on the 4.x branch
```
git clone https://github.com/tripal/tripal.git tripal-1404-chadoTestImprovements
cd tripal-1404-chadoTestImprovements
git checkout 4.x
docker run --publish=80:80 --name=tripal1404 -tid --volume=$(pwd):/var/www/drupal/web/modules/contrib/tripal tripalproject/tripaldocker:latest
docker exec tripal1404 service postgresql restart
```
2. Install a new chado instance and alter it using the following commands. THIS WILL ALTER THE CHADO SCHEMA SO ONLY DO ON A FRESH TRIPAL DOCKER CONTAINER.
```
docker exec -it tripal1404 drush trp-install-chado --schema-name=testupgradechado
docker exec -it tripal1404 drush sql:cli
```
```sql
DROP TABLE testupgradechado.analysis CASCADE;
ALTER TABLE testupgradechado.project DROP COLUMN description;
ALTER TABLE testupgradechado.feature ADD COLUMN testsum INT;
UPDATE testupgradechado.chadoprop SET value='1.1';
UPDATE chado_installations SET version=1.1 WHERE schema_name='testupgradechado';
```
4. Go to Admin > Tripal > Data Storage > Chado > Chado Schemas. You should see something like the following screenshot.
<img width="1089" alt="Screenshot 2023-12-01 at 4 12 53 PM" src="https://github.com/tripal/tripal/assets/1566301/01aa1003-94c2-4387-a9bf-c2c62230502c">

5. Click the "Upgrade to 1.3" button beside the testupgradechado schema under operations.
6. In the next form, enter the file name `/var/www/drupal/web/upgrade_test.sql` and click "Upgrade"
<img width="1088" alt="Screenshot 2023-12-01 at 4 14 33 PM" src="https://github.com/tripal/tripal/assets/1566301/eff88869-9027-4c5d-bc95-c74bcae26fa1">
7. Run the tripal job on the command line

```
docker exec tripal1404 drush trp-run-jobs --username=drupaladmin --root=/var/www/drupal/web
```
8. Copy the file produced to your local computer for comparison later.

```
docker cp tripal1404:/var/www/drupal/web/upgrade_test.sql upgradeTest_4.x.sql
docker exec tripal1404 rm /var/www/drupal/web/upgrade_test.sql 
```
9. Switch to this PR branch and clear the cache on the docker container.
```
git checkout tv4g0-issue1404-improveMaintainability-ChadoUpgraderComplexity
docker exec tripal1404 drush cr
```
10. Now repeat steps 4-7 above on this branch.
11. Then copy this upgrade.sql file to your local computer as well.
```
docker cp tripal1404:/var/www/drupal/web/upgrade_test.sql upgradeTest_1404.sql
```
12. Finally compare the two files to show that the changes in this PR did not change anything at all in terms of the resulting file.
```
diff upgradeTest_4.x.sql upgradeTest_1404.sql 
```
You should expect to see only the following difference between the files:
```diff
8136c8136
< UPDATE "chado_installations" SET version = '1.3', created = '1701468940', updated = '1701468940' WHERE schema_name = 'testupgradechado\';
---
> UPDATE "chado_installations" SET version = '1.3', created = '1701471602', updated = '1701471602' WHERE schema_name = 'testupgradechado';
```
